### PR TITLE
[C] Color Tool cached image callbacks

### PR DESCRIPTION
**Pull request by alexgoff**
JIRA: https://jira.lsstcorp.org/browse/EPO-9039

## What this change does

Fixed `ColorTool` not performing image load callbacks by refactoring to read `complete` attribute on images. Additionally moved all canvases into a single hook, no more headless components.

## Notes for reviewers

Can be tested in Storybook

Include an indication of how detailed a review you want on a 1-10 scale.

- 3

## Testing

Reload page with cache enabled and see that images complete loading. 

Check other functionality like applying filters and increasing color intensity.

### CODE REVIEW

File: packages/epo-widget-lib/src/widgets/ColorTool/ImageComposite/ImageComposite.tsx [hunk 1]
```diff
@@ -32,71 +33,49 @@ const ImageComposite = forwardRef<
...
     if (selectedObjectName !== prevObject) {
       setPrevObject(selectedObjectName);
-      setImagesLoaded(0);
     }
 
-    const onLoadCallback = useCallback(
-      () => setImagesLoaded((count) => count + 1),
-      [imagesLoaded, filters]
-    );
+    const [imgElements, loading] = useFilteredImages({
```
Comment (alexgoff) [inline packages/epo-widget-lib/src/widgets/ColorTool/ImageComposite/ImageComposite.tsx:48]: finally figured out how to put all of this into a simple hook

File: packages/epo-widget-lib/src/widgets/ColorTool/ImageComposite/ImageComposite.tsx [hunk 2]
```diff
@@ -32,71 +33,49 @@ const ImageComposite = forwardRef<
...
-        )}
-        <Styled.LoadingContainer
-          style={{ "--loading-opacity": isLoading ? 0 : 1 }}
-        >
-          {filters &&
-            filters.map((filter, i) => {
-              const { label, image, color, brightness } = filter;
-
-              return (
-                <OffscreenFilter
```
Comment (alexgoff) [inline packages/epo-widget-lib/src/widgets/ColorTool/ImageComposite/ImageComposite.tsx:77]: no more weird headless components, woohoo

File: packages/epo-widget-lib/src/widgets/ColorTool/hooks/useFilteredImages.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}: useFilteredImagesProps): [Array<HTMLCanvasElement>, boolean] => {
+  const previousImages = usePrevious(images);
+  const [imagesLoaded, setImagesLoaded] = useState(0);
+  const [imgElements, setImages] = useState<Array<HTMLImageElement>>();
+
+  const anyImageIncomplete = imgElements?.some(({ complete }) => !complete);
+
+  const loading = imagesLoaded !== images.length && anyImageIncomplete;
+
+  useLayoutEffect(() => {
```
Comment (alexgoff) [inline packages/epo-widget-lib/src/widgets/ColorTool/hooks/useFilteredImages.tsx:25]: trying to figure out how to read cached loading forced a rethink of a this whole setup. All of the images are now loaded as `Image` blobs that are put into an array in the current state, so they do not reload whenever props change.

File: packages/epo-widget-lib/src/widgets/ColorTool/hooks/useFilteredImages.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+          };
+          img.src = url;
+
+          return img;
+        })
+      );
+    }
+  }, [images]);
+
+  if (loading) {
```
Comment (alexgoff) [inline packages/epo-widget-lib/src/widgets/ColorTool/hooks/useFilteredImages.tsx:42]: we determine loading status for all of the items, and return nothing while loading is happening

**Review by blnkt — APPROVED** [review]
Approved, but you have some tests failing.  Possibly the packages won't build until you address.  If they do build with broken tests, I think you can merge in this work now and update the tests at a later date.  Whichever you prefer.

**Pull request merged**